storage: improve store visitor

tbg opened the pull request:
The previous code was using the Store btree to visit the Replicas in
a particular order. It turns out that order was never necessary (and perhaps
even counter-productive).

More importantly, since the btree version needed to lock each Replica mutex
while also holding the Store lock, Store.mu was as contended as the worst
Replica.mu, which is disastrous. The new code simply copies the Replicas
out from under the Store lock without invoking any methods on them, which
should completely avoid that problem.

Prompted by #9749, which showcased the above problem.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9771)

<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOLnAQjKrVSGv5AHiN:-KTOLnAQjKrVSGv5AHiO:b-uy9khy) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L243)):*
> ```Go
> 		repl.mu.Lock()
> 		destroyed := repl.mu.destroyed
> 		initialized := repl.isInitializedLocked()
> ```

where'd this come from?

---

*[storage/store.go, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOMaqSS1r392tmHRX_:-KTOMaqSS1r392tmHRXa:bejq15u) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L252)):*
> ```Go
> }
> 
> // TODO(tschottdorf): this method has highly doubtful semantics.
> ```

FYI this method has exactly one caller (scanner.go). Might be good to inline it over there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-:-KTOMotvp8XdPFwBM8de:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Might also be good to test that visiting doesn't try to lock the replica.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-:-KTOMsOnuxAUlgiac6YL:b-lmrm72)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 246 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTP_CaBkGt0YGRqYQXY:-KTP_CaBkGt0YGRqYQXZ:beypgfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d94915de34e6d146b6d3445d95dcca28d7ded19a/storage/store.go#L246)):*
> ```Go
> 	rs.store.mu.Unlock()
> 
> 	if rs.shuffle {
> ```

Perhaps we should leave this in. I believe the randomized iterations of Go maps isn't guaranteed. Is the shuffling causing a problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: improve store visitor

The previous code was using the Store btree to visit the Replicas in
a particular order. It turns out that order was never necessary (and perhaps
even counter-productive).

More importantly, since the btree version needed to lock each Replica mutex
while also holding the Store lock, Store.mu was as contended as the worst
Replica.mu, which is disastrous. The new code simply copies the Replicas
out from under the Store lock without invoking any methods on them, which
should completely avoid that problem.

Prompted by #9749, which showcased the above problem.

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 243 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOLnAQjKrVSGv5AHiN:-KTQBn6jJu49oETQOqSz:b-ea96sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> where'd this come from?
</details>
Uninitialized replicas are not in the btree but are in the replicas map.

---

*[storage/store.go, line 246 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTP_CaBkGt0YGRqYQXY:-KTQBha-kYxy05H3jngQ:bgtjjwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d94915de34e6d146b6d3445d95dcca28d7ded19a/storage/store.go#L246)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Perhaps we should leave this in. I believe the randomized iterations of Go maps isn't guaranteed. Is the shuffling causing a problem?
</details>
I doubt it but I agree that we shouldn't rely on map iteration order. Reinstated.

---

*[storage/store.go, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOMaqSS1r392tmHRX_:-KTQBuP2uLTi_eQOzzyf:bfqkynu) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> FYI this method has exactly one caller (scanner.go). Might be good to inline it over there.
</details>
It's also used in tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-:-KTQD149H-oTJvxgMNIU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOMaqSS1r392tmHRX_:-KTQG6Ly5nV6GDKxNiMy:b-jkza6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L252)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It's also used in tests.
</details>
I thought those tests were removed in this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 252 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771#-KTOMaqSS1r392tmHRX_:-KTQGS3TkTa0ByRRPVW_:b-91yqk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/03ecfdbcabc1357e5b48c8833a8f32c54e6e3310/storage/store.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I thought those tests were removed in this PR?
</details>
14:33 $ git grep EstimatedCount
storage/scanner.go:     // EstimatedCount returns the number of replicas estimated to remain
storage/scanner.go:     EstimatedCount() int
storage/scanner.go:     count := rs.replicas.EstimatedCount()
storage/scanner_test.go:func (rs *testRangeSet) EstimatedCount() int {
storage/store.go:func (rs *storeReplicaVisitor) EstimatedCount() int {
storage/store_test.go:          if ec := ranges.EstimatedCount(); ec != 10 {
storage/store_test.go:                  if ec := ranges.EstimatedCount(); ec != 10-i {
storage/store_test.go:          if ec := ranges.EstimatedCount(); ec != 10 {

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9771)*
<!-- Sent from Reviewable.io -->
